pkg/cmd/server/kubernetes/node_config.go

```diff
@@ -111,29 +84,85 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
 		dockerExecHandler = &dockertools.NsenterExecHandler{}
 	}
 
-	config := &NodeConfig{
-		NodeHost:    options.NodeName,
-		BindAddress: options.ServingInfo.BindAddress,
+	kubeAddress, kubePortStr, err := net.SplitHostPort(options.ServingInfo.BindAddress)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse node address: %v", err)
+	}
+	kubePort, err := strconv.Atoi(kubePortStr)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse node port: %v", err)
+	}
 
-		TLS:             configapi.UseTLS(options.ServingInfo),
-		KubeletCertFile: options.ServingInfo.ServerCert.CertFile,
-		KubeletKeyFile:  options.ServingInfo.ServerCert.KeyFile,
-		ClientCAs:       clientCAs,
+	address := util.IP{}
+	if err := address.Set(kubeAddress); err != nil {
+		return nil, err
+	}
 
-		ClusterDomain: options.DNSDomain,
-		ClusterDNS:    dnsIP,
+	// declare the OpenShift defaults from config
+	server := kapp.NewKubeletServer()
+	server.Config = path
+	server.RootDirectory = options.VolumeDirectory
+	server.HostnameOverride = options.NodeName
+	server.AllowPrivileged = true
+	server.RegisterNode = true
+	server.Address = address
+	server.Port = uint(kubePort)
+	server.ReadOnlyPort = 0 // no read only access
+	server.ClusterDNS = util.IP(dnsIP)
+	server.ClusterDomain = options.DNSDomain
+	server.NetworkPluginName = options.NetworkPluginName
+	server.HostNetworkSources = strings.Join([]string{kubelet.ApiserverSource, kubelet.FileSource}, ",")
+	server.HTTPCheckFrequency = 0 // no remote HTTP pod creation access
+	server.FileCheckFrequency = time.Duration(fileCheckInterval) * time.Second
+	server.PodInfraContainerImage = imageTemplate.ExpandOrDie("pod")
+
+	// prevents kube from generating certs
+	server.TLSCertFile = options.ServingInfo.ServerCert.CertFile
+	server.TLSPrivateKeyFile = options.ServingInfo.ServerCert.KeyFile
+
+	// resolve extended arguments
+	if err := cmdflags.Resolve(options.KubeletArguments, server.AddFlags); len(err) > 0 {
```

Comment:
Two more things: 
1. either way the override happens, document it the field comment
1. this resolution can be done during config validation and warnings (or errors if you're going to fail) can be produced